File: KernelLsh/C++/kernel_lsh.cc
// Implement Locality-Sensitive Binary Codesfrom Shift-Invariant Kernels NIPS(2009)
// http://www.robots.ox.ac.uk/~vgg/rg/papers/binarycodes.pdf
// cos(w*x + b) + t
// Author: Yi Yang (yangyi_bupt@126.com)
//

#include "kernel_lsh.h"

#include <cmath>

namespace {
template<class T>
inline int32_t sign(T val) {
    return val >= 0 ? 1:0;
}

double Dot(const feature_node *px, const feature_node *py) {
    double sum = 0;
    while (px->index != -1 && py->index != -1) {
        if (px->index == py->index) {
            sum += px->value * py->value;
            ++px;
            ++py;
        } else {
            px->index > py->index ? ++py:++px;
        }
    }
    return sum;
}

} // namespace

KLSH::KLSH(const uint32_t& max_feat_num, const int& hash_bit_num, const float& sigma)
        : m_max_feat_num(max_feat_num), m_hash_bit_num(hash_bit_num) {

    m_normal.reset(new std::normal_distribution<double>(0.0, sigma * sigma));
    m_uniform_1.reset(new std::uniform_real_distribution<double>(-1.0, 1.0));
    m_uniform_pi.reset(new std::uniform_real_distribution<double>(0.0, 2.0 * M_PI));

    m_project_w.reset(new scoped_array<feature_node>[m_hash_bit_num]);
    m_project_t.reset(new double[m_hash_bit_num]);
    m_project_b.reset(new double[m_hash_bit_num]);

    for (uint32_t i = 0; i < m_hash_bit_num; ++i) {
        // init project_t, project_b
        m_project_t[i] = (*m_uniform_1)(eng);
        m_project_b[i] = (*m_uniform_pi)(eng);
        // init project_w (m_max_feat_num + 1, 1 more space for index -1)
        m_project_w[i].reset(new feature_node[m_max_feat_num + 1]);
        for (uint32_t j = 0; j < m_max_feat_num; ++j) {
            // svm index begin from 1
            m_project_w[i][j].index = j + 1;
            m_project_w[i][j].value = (*m_normal)(eng);
        }
        m_project_w[i][m_max_feat_num].index = -1;
    }
}

uint64_t KLSH::GetHashValue(const feature_node* pfeat) {
    float sum = 0;
    uint64_t hash_value = 0;
    for (uint32_t i = 0; i < m_hash_bit_num; ++i) {
        sum = static_cast<float>(Dot(pfeat, m_project_w[i].get()) + m_project_b[i]);
        hash_value |= static_cast<uint64_t>(sign(std::cos(sum) + m_project_t[i])) << i;
    }
    return hash_value;
}

File: KernelLsh/Java/src/main/java/pers/yy/algorithm/lsh/KernelLsh.java
package pers.yy.algorithm.lsh;

import org.apache.commons.math3.random.RandomDataGenerator;
import org.apache.mahout.math.SequentialAccessSparseVector;
import org.apache.mahout.math.Vector;

/**
 * Implement Locality-Sensitive Binary Codes from Shift-Invariant Kernels NIPS(2009)
 * http://www.robots.ox.ac.uk/~vgg/rg/papers/binarycodes.pdf
 * cos(w*x + b) + t 
 * 
 * TODO(yiyang): consider to use a run-time gen random method to save space
 *
 * @author Yi Yang (yangyi_bupt@126.com)
 */

public class KernelLsh {
    protected SequentialAccessSparseVector[] projectW;
    protected SequentialAccessSparseVector projectT;
    protected SequentialAccessSparseVector projectB;

    protected int cardinality;
    protected int hashBitNum;

    protected RandomDataGenerator gen = new RandomDataGenerator();
    protected static double pi2 = 2.0 * Math.PI;
    protected double sigmaSquare = -1;
    protected double eps = 0.0001;

    public KernelLsh(int cardinality, int hashBitNum, float sigma) {
        this.cardinality = cardinality;
        this.hashBitNum = hashBitNum;

        projectW = new SequentialAccessSparseVector[hashBitNum];
        projectB = new SequentialAccessSparseVector(hashBitNum);
        projectT = new SequentialAccessSparseVector(hashBitNum);

        RandomDataGenerator gen = new RandomDataGenerator();
        gen.reSeed(1000); // 1000 does not means anything, only to make a fixed sequence

        sigmaSquare = sigma * sigma;
        double tmpGenRes = 0;
        for (int i = 0; i < hashBitNum; ++i) {
            projectT.setQuick(i, gen.nextUniform(-1, 1));
            projectB.setQuick(i, gen.nextUniform(0.0, pi2));
            projectW[i] = new SequentialAccessSparseVector(cardinality);

            for (int j = 0; j < cardinality; ++j) {
                tmpGenRes = gen.nextGaussian(0.0, sigmaSquare);
                // truncate 0
                if (Math.abs(tmpGenRes) > eps) {
                    projectW[i].setQuick(j, tmpGenRes);
                }
            }
        }
    }
    
    public long GetHashValue(final Vector feat) {
        double sum = 0;
        long hash_value = 0;
        // TODO(yiyang): use matrix operations
        for (int i = 0; i < hashBitNum; ++i) {
            sum = feat.dot(projectW[i]) + projectB.get(i);
            if (Math.signum(Math.cos(sum) + projectT.get(i)) >= 0) {
                hash_value |= (long) 1 << i;
            }
        }                                                                                               
        return hash_value;
    }
}
